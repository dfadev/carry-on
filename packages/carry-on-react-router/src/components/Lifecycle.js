import React from "react";

class Lifecycle extends React.Component {
  componentDidMount() {
    const { onMount } = this.props;
    if (onMount) onMount.call(this, this);
  }

  componentDidUpdate(prevProps) {
    const { onUpdate } = this.props;
    if (onUpdate) onUpdate.call(this, this, prevProps);
  }

  componentWillUnmount() {
    const { onUnmount } = this.props;
    if (onUnmount) onUnmount.call(this, this);
  }

  render() {
    return null;
  }
}

export default Lifecycle;
