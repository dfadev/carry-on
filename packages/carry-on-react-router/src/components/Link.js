import React from "react";
import { carryOn } from "carry-on-react";
import { getIn } from "carry-on-utils";

const Link = carryOn(
  { id: "Link" },
  (
    {
      history = "app.history",
      innerRef,
      replace,
      to,
      onClick,
      target,
      force,
      ...rest
    },
    state
  ) => {
    const { handleClick, getHref } = getIn(state, history);

    return (
      <a
        {...rest}
        onClick={handleClick({ replace, to, onClick, target, force })}
        href={getHref(to)}
        ref={innerRef}
      />
    );
  }
);

export default Link;
