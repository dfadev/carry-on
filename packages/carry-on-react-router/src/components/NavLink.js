import React from "react";
import { carryOn } from "carry-on-react";
import { getIn } from "carry-on-utils";

function joinClassnames(...classnames) {
  return classnames.filter(i => i).join(" ");
}

const escapedRE = /([.+*?=^!:${}()[\]|/\\])/g;

const NavLink = carryOn(
  { id: "NavLink" },
  (
    {
      history = "app.history",
      "aria-current": ariaCurrent = "page",
      activeClassName = "active",
      activeStyle,
      className: classNameProp,
      exact,
      isActive: isActiveProp,
      location: loc,
      style: styleProp,
      strict,
      sensitive,
      to,
      replace,
      onClick,
      target,
      innerRef,
      ...rest
    },
    state
  ) => {
    const hist = getIn(state, history);
    const location = loc || hist.location;
    const path = typeof to === "object" ? to.pathname : to;
    const escapedPath = path && path.replace(escapedRE, "\\$1");
    const match = hist.matchPath({
      path: escapedPath,
      exact,
      strict,
      sensitive
    });
    const isActive = !!(isActiveProp ? isActiveProp(match, location) : match);
    const className = isActive
      ? joinClassnames(classNameProp, activeClassName)
      : classNameProp;
    const style = isActive ? { ...styleProp, ...activeStyle } : styleProp;
    const { handleClick, getHref } = hist;

    /* eslint-disable jsx-a11y/anchor-has-content */
    return (
      <a
        {...rest}
        onClick={handleClick({ replace, to, onClick, target })}
        href={getHref(to)}
        ref={innerRef}
        style={style}
        className={className}
        aria-current={(isActive && ariaCurrent) || null}
      />
    );
  }
);

export default NavLink;
