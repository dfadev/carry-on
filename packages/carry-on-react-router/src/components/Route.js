import React from "react";
import { carryOn } from "carry-on-react";
import { getIn } from "carry-on-utils";

function isEmptyChildren(children) {
  return React.Children.count(children) === 0;
}

const Route = carryOn((props, state) => {
  const {
    history = "app.history",
    location: loc,
    computedMatch,
    path,
    children: childs,
    component,
    render
  } = props;

  const hist = getIn(state, history);
  const location = loc || hist.location;

  let match;
  if (computedMatch) match = computedMatch;
  else if (path) match = hist.matchPath(props);
  //else match = hist.match; // ???

  const renderProps = { location, match };

  let children = childs;
  if (Array.isArray(children) && children.length === 0)
    children = null;

  if (typeof children === "function") {
    children = children(renderProps);
    if (children === undefined) children = null;
  }

  if (children && !isEmptyChildren(children)) return children;
  if (match)
    if (component) return React.createElement(component, renderProps);
    else if (render) return render(renderProps);

  return null;
});

Route.displayName = "Route";

export default Route;
