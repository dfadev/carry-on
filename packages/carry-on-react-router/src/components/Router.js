/* eslint-disable max-classes-per-file */
import React, { useMemo } from "react";
import {
  withStore,
  State,
  Register,
  Middleware,
  OnUnmount,
  Render
} from "carry-on-react";
import {
  createBrowserHistory,
  createHashHistory,
  createMemoryHistory
} from "history";
import createStaticHistory from "../createStaticHistory";
import router from "../router";

function GenericRouter(createHistory = createBrowserHistory) {
  return withStore(({ children = null, path = "app.history", ...props }) => {
    const rtr = useMemo(() => router(createHistory(props), path), [path]);

    return (
      <State {...props}>
        <Register>{rtr.state}</Register>
        <Middleware>{rtr.middleware}</Middleware>
        <OnUnmount>
          {state =>
            state &&
            state.app &&
            state.app.history &&
            state.app.history.unlisten &&
            state.app.history.unlisten()
          }
        </OnUnmount>
        <Render>{() => children || null}</Render>
      </State>
    );
  });
}

export const Router = GenericRouter(props => props.history);
export const MemoryRouter = GenericRouter(createMemoryHistory);
export const BrowserRouter = GenericRouter(createBrowserHistory);
export const HashRouter = GenericRouter(createHashHistory);
export const StaticRouter = GenericRouter(createStaticHistory);
