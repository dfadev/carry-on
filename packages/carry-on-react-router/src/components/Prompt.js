import React from "react";
import { carryOn } from "carry-on-react";
import { getIn } from "carry-on-utils";
import Lifecycle from "./Lifecycle";

/**
 * The public API for prompting the user before navigating away from a screen.
 */
const Prompt = carryOn(
  { id: "Prompt" },
  ({ history = "app.history", message, when = true }, state) => {
    const hist = getIn(state, history);

    if (!when || hist.staticContext) return null;

    const method = hist.block;

    return (
      <Lifecycle
        onMount={self => {
          self.release = method(message);
        }}
        onUpdate={(self, prevProps) => {
          if (prevProps.message !== message) {
            self.release();
            self.release = method(message);
          }
        }}
        onUnmount={self => {
          self.release();
        }}
        message={message}
      />
    );
  }
);

export default Prompt;
