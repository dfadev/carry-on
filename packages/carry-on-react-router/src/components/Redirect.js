import React from "react";
import { createLocation, locationsAreEqual } from "history";
import { carryOn } from "carry-on-react";
import { getIn } from "carry-on-utils";

import Lifecycle from "./Lifecycle";
import generatePath from "../generatePath";

const Redirect = carryOn(
  { id: "Redirect" },
  ({ history = "app.history", computedMatch, to, push = false }, state) => {
    const hist = getIn(state, history);
    const method = push ? hist.push : hist.replace;

    /* eslint-disable no-nested-ternary */
    const location = createLocation(
      computedMatch
        ? typeof to === "string"
          ? generatePath(to, computedMatch.params)
          : {
              ...to,
              pathname: generatePath(to.pathname, computedMatch.params)
            }
        : to
    );

    // When rendering in a static context, set the new location immediately.
    if (hist.staticContext) {
      method(location);
      return null;
    }

    return (
      <Lifecycle
        onMount={() => {
          method(location);
        }}
        onUpdate={(_, prevProps) => {
          if (!locationsAreEqual(prevProps.to, location)) {
            method(location);
          }
        }}
        to={to}
      />
    );
  }
);

export default Redirect;
