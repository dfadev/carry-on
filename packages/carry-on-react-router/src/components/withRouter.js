/* eslint-disable react/jsx-props-no-spreading */
import React from "react";
import hoistStatics from "hoist-non-react-statics";
import Route from "./Route";

function withRouter(Component) {
  const C = props => {
    const { wrappedComponentRef, ...remainingProps } = props;

    return (
      <Route>
        {routeComponentProps => (
          <Component
            {...remainingProps}
            {...routeComponentProps}
            ref={wrappedComponentRef}
          />
        )}
      </Route>
    );
  };

  C.displayName = `withRouter(${Component.displayName || Component.name})`;
  C.WrappedComponent = Component;

  return hoistStatics(C, Component);
}

export default withRouter;
