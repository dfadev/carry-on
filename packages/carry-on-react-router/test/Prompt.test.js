import React from "react";
import { render } from "react-testing-library";
import { createMemoryHistory as createHistory } from "history";
import { MemoryRouter, Router } from "../src/components/Router";
import NavLink from "../src/components/NavLink";
import withRouter from "../src/components/withRouter";
import Prompt from "../src/components/Prompt";
import { initStores } from "carry-on-store";

describe("A <Prompt>", () => {
  afterEach(() => {
    initStores();
  });

  describe("with no message", () => {
    it("renders", () => {
      expect(
        render(
          <MemoryRouter>
            <Prompt />
          </MemoryRouter>
        ).asFragment()
      ).toMatchSnapshot();
    });
  });

  it("calls getUserConfirmation with the prompt message", () => {
    const getUserConfirmation = jest.fn((message, callback) => {
      callback(false);
    });

    const history = createHistory({
      getUserConfirmation: getUserConfirmation
    });

    render(
      <Router history={history}>
        <Prompt message="Are you sure?" />
      </Router>
    );

    history.push("/somewhere");

    expect(getUserConfirmation).toHaveBeenCalledWith(
      expect.stringMatching("Are you sure?"),
      expect.any(Function)
    );
  });

  describe("with when=false", () => {
    it("does not call getUserConfirmation", () => {
      const getUserConfirmation = jest.fn((message, callback) => {
        callback(false);
      });

      const history = createHistory({
        getUserConfirmation: getUserConfirmation
      });

      render(
        <Router history={history}>
          <Prompt message="Are you sure?" when={false} />
        </Router>
      );

      history.push("/somewhere");

      expect(getUserConfirmation).not.toHaveBeenCalled();
    });
  });
});
