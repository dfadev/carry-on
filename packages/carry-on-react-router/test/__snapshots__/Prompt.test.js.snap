// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`A <Prompt> with no message renders 1`] = `<DocumentFragment />`;
